In [ ]:
import kagglehub

# Download latest version
path = kagglehub.dataset_download("maxhorowitz/nflplaybyplay2009to2016")

print("Path to dataset files:", path)

100%|██████████| 274M/274M [00:03<00:00, 92.9MB/s]

Extracting files...


Path to dataset files: /root/.cache/kagglehub/datasets/maxhorowitz/nflplaybyplay2009to2016/versions/6


In [ ]:
import pandas as pd
import numpy as np

df = pd.read_csv(path + "/NFL Play by Play 2009-2017 (v4).csv")
df.head()

# Set seed for NumPy's random number generator
np.random.seed(42)

# Set options to display all columns and all rows!
pd.set_option('display.max_columns', None)  # No column truncation. You will see all columns!
pd.set_option('display.max_rows', None)     # No row truncation. You will see all rows!



<ipython-input-3-0473edf1c15c>:4: DtypeWarning: Columns (25,51) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv(path + "/NFL Play by Play 2009-2017 (v4).csv")


Random seed

In many data science and machine learning tasks, we use functions or algorithms that involve randomness. A prime example is splitting data
into training and testing sets, or initializing weights in a neural network. To ensure that we get the same results every time we run our
code, even with these random elements, we use the concept of "setting the seed."

Why is it important?

* Reproducibility: Setting the seed allows us to reproduce the exact same results when we re-run our code. This is crucial for debugging,
sharing results, and ensuring consistency in experiments.

* Comparability: When comparing different models or algorithms, it's essential to ensure that the randomness factor is controlled, making
the comparisons fair and meaningful.

* Collaboration: When working in teams, setting the seed helps everyone obtain the same results, facilitating collaboration and avoiding
discrepancies.


In [ ]:
df.head()

Date      GameID  Drive  qtr  down   time  TimeUnder  TimeSecs  \
0  2009-09-10  2009091000      1    1   NaN  15:00         15    3600.0   
1  2009-09-10  2009091000      1    1   1.0  14:53         15    3593.0   
2  2009-09-10  2009091000      1    1   2.0  14:16         15    3556.0   
3  2009-09-10  2009091000      1    1   3.0  13:35         14    3515.0   
4  2009-09-10  2009091000      1    1   4.0  13:27         14    3507.0   

   PlayTimeDiff SideofField  yrdln  yrdline100  ydstogo  ydsnet  GoalToGo  \
0           0.0         TEN   30.0        30.0        0       0       0.0   
1           7.0         PIT   42.0        58.0       10       5       0.0   
2          37.0         PIT   47.0        53.0        5       2       0.0   
3          41.0         PIT   44.0        56.0        8       2       0.0   
4           8.0         PIT   44.0        56.0        8       2       0.0   

   FirstDown posteam DefensiveTeam  \
0        NaN     PIT           TEN   
1        0.0     PIT           TEN   
2        0.0     PIT           TEN   
3        0.0     PIT           TEN   
4        1.0     PIT           TEN   

                                                desc  PlayAttempted  \
0  R.Bironas kicks 67 yards from TEN 30 to PIT 3....              1   
1  (14:53) B.Roethlisberger pass short left to H....              1   
2  (14:16) W.Parker right end to PIT 44 for -3 ya...              1   
3  (13:35) (Shotgun) B.Roethlisberger pass incomp...              1   
4  (13:27) (Punt formation) D.Sepulveda punts 54 ...              1   

   Yards.Gained  sp  Touchdown ExPointResult TwoPointConv DefTwoPoint  Safety  \
0            39   0          0           NaN          NaN         NaN       0   
1             5   0          0           NaN          NaN         NaN       0   
2            -3   0          0           NaN          NaN         NaN       0   
3             0   0          0           NaN          NaN         NaN       0   
4             0   0          0           NaN          NaN         NaN       0   

   Onsidekick PuntResult PlayType            Passer   Passer_ID  PassAttempt  \
0           0        NaN  Kickoff               NaN         NaN            0   
1           0        NaN     Pass  B.Roethlisberger  00-0022924            1   
2           0        NaN      Run               NaN         NaN            0   
3           0        NaN     Pass  B.Roethlisberger  00-0022924            1   
4           0      Clean     Punt               NaN         NaN            0   

       PassOutcome PassLength  AirYards  YardsAfterCatch  QBHit PassLocation  \
0              NaN        NaN         0                0      0          NaN   
1         Complete      Short        -3                8      0         left   
2              NaN        NaN         0                0      0          NaN   
3  Incomplete Pass       Deep        34                0      0        right   
4              NaN        NaN         0                0      0          NaN   

   InterceptionThrown Interceptor    Rusher   Rusher_ID  RushAttempt  \
0                   0         NaN       NaN         NaN            0   
1                   0         NaN       NaN         NaN            0   
2                   0         NaN  W.Parker  00-0022250            1   
3                   0         NaN       NaN         NaN            0   
4                   0         NaN       NaN         NaN            0   

  RunLocation RunGap   Receiver Receiver_ID  Reception ReturnResult Returner  \
0         NaN    NaN        NaN         NaN          0          NaN  S.Logan   
1         NaN    NaN     H.Ward  00-0017162          1          NaN      NaN   
2       right    end        NaN         NaN          0          NaN      NaN   
3         NaN    NaN  M.Wallace  00-0026901          0          NaN      NaN   
4         NaN    NaN        NaN         NaN          0          NaN      NaN   

  BlockingPlayer   Tackler1 Tackler2 FieldGoalResult  FieldGoalDistance  \
0

In [ ]:
df.shape

(407688, 102)

In [ ]:
df.columns

Index(['Date', 'GameID', 'Drive', 'qtr', 'down', 'time', 'TimeUnder',
       'TimeSecs', 'PlayTimeDiff', 'SideofField',
       ...
       'yacEPA', 'Home_WP_pre', 'Away_WP_pre', 'Home_WP_post', 'Away_WP_post',
       'Win_Prob', 'WPA', 'airWPA', 'yacWPA', 'Season'],
      dtype='object', length=102)

In [ ]:
df.dtypes

Date                            object
GameID                           int64
Drive                            int64
qtr                              int64
down                           float64
time                            object
TimeUnder                        int64
TimeSecs                       float64
PlayTimeDiff                   float64
SideofField                     object
yrdln                          float64
yrdline100                     float64
ydstogo                          int64
ydsnet                           int64
GoalToGo                       float64
FirstDown                      float64
posteam                         object
DefensiveTeam                   object
desc                            object
PlayAttempted                    int64
Yards.Gained                     int64
sp                               int64
Touchdown                        int64
ExPointResult                   object
TwoPointConv                    object
DefTwoPoint                     object
Safety                           int64
Onsidekick                       int64
PuntResult                      object
PlayType                        object
Passer                          object
Passer_ID                       object
PassAttempt                      int64
PassOutcome                     object
PassLength                      object
AirYards                         int64
YardsAfterCatch                  int64
QBHit                            int64
PassLocation                    object
InterceptionThrown               int64
Interceptor                     object
Rusher                          object
Rusher_ID                       object
RushAttempt                      int64
RunLocation                     object
RunGap                          object
Receiver                        object
Receiver_ID                     object
Reception                        int64
ReturnResult                    object
Returner                        object
BlockingPlayer                  object
Tackler1                        object
Tackler2                        object
FieldGoalResult                 object
FieldGoalDistance              float64
Fumble                           int64
RecFumbTeam                     object
RecFumbPlayer                   object
Sack                             int64
Challenge.Replay                 int64
ChalReplayResult                object
Accepted.Penalty                 int64
PenalizedTeam                   object
PenaltyType                     object
PenalizedPlayer                 object
Penalty.Yards                    int64
PosTeamScore                   float64
DefTeamScore                   float64
ScoreDiff                      float64
AbsScoreDiff                   float64
HomeTeam                        object
AwayTeam                        object
Timeout_Indicator                int64
Timeout_Team                    object
posteam_timeouts_pre             int64
HomeTimeouts_Remaining_Pre       int64
AwayTimeouts_Remaining_Pre       int64
HomeTimeouts_Remaining_Post      int64
AwayTimeouts_Remaining_Post      int64
No_Score_Prob                  float64
Opp_Field_Goal_Prob            float64
Opp_Safety_Prob                float64
Opp_Touchdown_Prob             float64
Field_Goal_Prob                float64
Safety_Prob                    float64
Touchdown_Prob                 float64
ExPoint_Prob                   float64
TwoPoint_Prob                  float64
ExpPts                         float64
EPA                            float64
airEPA                         float64
yacEPA                         float64
Home_WP_pre                    float64
Away_WP_pre                    float64
Home_WP_post                   float64
Away_WP_post                   float64
Win_Prob                       float64
WPA                            float64
airWPA                         float64
yacWPA                         float64
Season                           int64
dtype: object

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 407688 entries, 0 to 407687
Columns: 102 entries, Date to Season
dtypes: float64(33), int64(31), object(38)
memory usage: 317.3+ MB


In [ ]:
# Get the number of missing data per column

missing_values_count = df.isnull().sum()

missing_values_count.sort_values(ascending=False)

DefTwoPoint                    407664
BlockingPlayer                 407571
TwoPointConv                   407083
ChalReplayResult               404286
RecFumbTeam                    403315
RecFumbPlayer                  403315
Interceptor                    403168
FieldGoalDistance              398740
FieldGoalResult                398629
ExPointResult                  397578
PenaltyType                    391261
Timeout_Team                   390885
ReturnResult                   389450
PuntResult                     385317
Returner                       382384
PenalizedPlayer                379403
PenalizedTeam                  378189
Tackler2                       357953
RunGap                         320260
RunLocation                    288178
Rusher                         287124
Rusher_ID                      283654
Receiver_ID                    250328
yacWPA                         248762
airWPA                         248501
yacEPA                         248498
airEPA                         248394
Receiver                       246127
PassLocation                   240520
PassLength                     240520
Passer                         240045
PassOutcome                    239506
Passer_ID                      237445
Tackler1                       166134
down                            61154
FirstDown                       28811
DefTeamScore                    26904
PosTeamScore                    26904
AbsScoreDiff                    26904
Home_WP_post                    26587
Away_WP_post                    26587
Win_Prob                        25009
DefensiveTeam                   24992
posteam                         24992
ScoreDiff                       24988
Away_WP_pre                     24954
Home_WP_pre                     24954
WPA                              5541
GoalToGo                          840
yrdline100                        840
yrdln                             840
SideofField                       528
PlayTimeDiff                      444
EPA                               369
TimeSecs                          224
time                              224
No_Score_Prob                     176
Opp_Field_Goal_Prob               176
Opp_Touchdown_Prob                176
Safety_Prob                       176
Touchdown_Prob                    176
Opp_Safety_Prob                   176
Field_Goal_Prob                   176
ExpPts                            176
desc                                2
AirYards                            0
TimeUnder                           0
RushAttempt                         0
qtr                                 0
Drive                               0
Touchdown                           0
sp                                  0
Yards.Gained                        0
PlayAttempted                       0
Fumble                              0
PassAttempt                         0
PlayType                            0
Onsidekick                          0
YardsAfterCatch                     0
Safety                              0
InterceptionThrown                  0
QBHit                               0
Reception                           0
Penalty.Yards                       0
Accepted.Penalty                    0
Sack                                0
AwayTimeouts_Remaining_Pre          0
HomeTimeouts_Remaining_Pre          0
posteam_timeouts_pre                0
ydsnet                              0
ydstogo                             0
HomeTeam                            0
AwayTeam                            0
Timeout_Indicator                   0
Challenge.Replay                    0
HomeTimeouts_Remaining_Post         0
AwayTimeouts_Remaining_Post         0
ExPoint_Prob                        0
TwoPoint_Prob                       0
GameID                              0
Date                                0
Season                              0
dtype: int64

missing_values_count = df.isnull().sum()

df.is null() converts the whole dataframe into a matrix of True and False. True is assigned when a missing values is identfied. Then sum() add all True or 1 (ones) per column.

In [ ]:
# Calculating the total number of missing values in the dataframe

total_missing = missing_values_count.sum()

print(f"{total_missing:,}")

11,505,187


In [ ]:
# Calculating the % of missing values in the dataframe

total_cells = np.product(df.shape)    # First, find the total number of cells in a dataframe

print(f"{total_cells:,}")

percent_missing = (total_missing / total_cells)

print(f"percentage of missing values: {percent_missing:.2%}")


41,584,176
percentage of missing values: 27.67%


**df.shape** produceds a tuple of the number rows and the number of columns (407688, 102). Then wrapped on **np.product**(407688, 102), it outputs the multiplication of both values. This is the total number of cells in the dataframe which is the denominator.

# **Approaches to handle missing values:**






# **Dropping missing values**
df.dropna() looks at every row in the dataFrame. If any cell within that row has a missing value (NaN), the entire row is deleted.

In [ ]:
df.dropna()

Empty DataFrame
Columns: [Date, GameID, Drive, qtr, down, time, TimeUnder, TimeSecs, PlayTimeDiff, SideofField, yrdln, yrdline100, ydstogo, ydsnet, GoalToGo, FirstDown, posteam, DefensiveTeam, desc, PlayAttempted, Yards.Gained, sp, Touchdown, ExPointResult, TwoPointConv, DefTwoPoint, Safety, Onsidekick, PuntResult, PlayType, Passer, Passer_ID, PassAttempt, PassOutcome, PassLength, AirYards, YardsAfterCatch, QBHit, PassLocation, InterceptionThrown, Interceptor, Rusher, Rusher_ID, RushAttempt, RunLocation, RunGap, Receiver, Receiver_ID, Reception, ReturnResult, Returner, BlockingPlayer, Tackler1, Tackler2, FieldGoalResult, FieldGoalDistance, Fumble, RecFumbTeam, RecFumbPlayer, Sack, Challenge.Replay, ChalReplayResult, Accepted.Penalty, PenalizedTeam, PenaltyType, PenalizedPlayer, Penalty.Yards, PosTeamScore, DefTeamScore, ScoreDiff, AbsScoreDiff, HomeTeam, AwayTeam, Timeout_Indicator, Timeout_Team, posteam_timeouts_pre, HomeTimeouts_Remaining_Pre, AwayTimeouts_Remaining_Pre, HomeTimeouts_Remaining_Post, AwayTimeouts_Remaining_Post, No_Score_Prob, Opp_Field_Goal_Prob, Opp_Safety_Prob, Opp_Touchdown_Prob, Field_Goal_Prob, Safety_Prob, Touchdown_Prob, ExPoint_Prob, TwoPoint_Prob, ExpPts, EPA, airEPA, yacEPA, Home_WP_pre, Away_WP_pre, Home_WP_post, Away_WP_post, Win_Prob, WPA, airWPA, ...]
Index: []

In this case, dropping all NaN of the dataframe is not a good idea as it has deleted all rows!

By setting axis=1 in dropna(), it operates on columns instead of rows (which is the default with axis=0). This removes columns with at least one missing value.

In [ ]:
columns_with_na_dropped = df.dropna(axis=1)   # remove all columns with at least one missing value

print(f"Updated number of columns: {columns_with_na_dropped.shape[1]}")
print(f"Original number of columns: {df.shape[1]}")


Updated number of columns: 37
Original number of columns: 102


In [ ]:
df_cleaned_all = df.dropna(axis=1, how='all')  # Remove only if all values in a column are NaN

print(f"Updated number of columns with all: {df_cleaned_all.shape[1]}")

Updated number of columns with all: 102


# **Filling in missing values automatically**

In [ ]:
# Replacing all NA's with 0

df.fillna(0).head()

Date      GameID  Drive  qtr  down   time  TimeUnder  TimeSecs  \
0  2009-09-10  2009091000      1    1   0.0  15:00         15    3600.0   
1  2009-09-10  2009091000      1    1   1.0  14:53         15    3593.0   
2  2009-09-10  2009091000      1    1   2.0  14:16         15    3556.0   
3  2009-09-10  2009091000      1    1   3.0  13:35         14    3515.0   
4  2009-09-10  2009091000      1    1   4.0  13:27         14    3507.0   

   PlayTimeDiff SideofField  yrdln  yrdline100  ydstogo  ydsnet  GoalToGo  \
0           0.0         TEN   30.0        30.0        0       0       0.0   
1           7.0         PIT   42.0        58.0       10       5       0.0   
2          37.0         PIT   47.0        53.0        5       2       0.0   
3          41.0         PIT   44.0        56.0        8       2       0.0   
4           8.0         PIT   44.0        56.0        8       2       0.0   

   FirstDown posteam DefensiveTeam  \
0        0.0     PIT           TEN   
1        0.0     PIT           TEN   
2        0.0     PIT           TEN   
3        0.0     PIT           TEN   
4        1.0     PIT           TEN   

                                                desc  PlayAttempted  \
0  R.Bironas kicks 67 yards from TEN 30 to PIT 3....              1   
1  (14:53) B.Roethlisberger pass short left to H....              1   
2  (14:16) W.Parker right end to PIT 44 for -3 ya...              1   
3  (13:35) (Shotgun) B.Roethlisberger pass incomp...              1   
4  (13:27) (Punt formation) D.Sepulveda punts 54 ...              1   

   Yards.Gained  sp  Touchdown ExPointResult TwoPointConv DefTwoPoint  Safety  \
0            39   0          0             0            0           0       0   
1             5   0          0             0            0           0       0   
2            -3   0          0             0            0           0       0   
3             0   0          0             0            0           0       0   
4             0   0          0             0            0           0       0   

   Onsidekick PuntResult PlayType            Passer   Passer_ID  PassAttempt  \
0           0          0  Kickoff                 0           0            0   
1           0          0     Pass  B.Roethlisberger  00-0022924            1   
2           0          0      Run                 0           0            0   
3           0          0     Pass  B.Roethlisberger  00-0022924            1   
4           0      Clean     Punt                 0           0            0   

       PassOutcome PassLength  AirYards  YardsAfterCatch  QBHit PassLocation  \
0                0          0         0                0      0            0   
1         Complete      Short        -3                8      0         left   
2                0          0         0                0      0            0   
3  Incomplete Pass       Deep        34                0      0        right   
4                0          0         0                0      0            0   

   InterceptionThrown Interceptor    Rusher   Rusher_ID  RushAttempt  \
0                   0           0         0           0            0   
1                   0           0         0           0            0   
2                   0           0  W.Parker  00-0022250            1   
3                   0           0         0           0            0   
4                   0           0         0           0            0   

  RunLocation RunGap   Receiver Receiver_ID  Reception ReturnResult Returner  \
0           0      0          0           0          0            0  S.Logan   
1           0      0     H.Ward  00-0017162          1            0        0   
2       right    end          0           0          0            0        0   
3           0      0  M.Wallace  00-0026901          0            0        0   
4           0      0          0           0          0            0        0   

  BlockingPlayer   Tackler1 Tackler2 FieldGoalResult  FieldGoalDistance  \
0

# **Replacing missing values with other metrics:**

```
df['column_name'] = df['column_name'].fillna(df['column_name'].median())

df['column_name'] = df['column_name'].fillna(df['column_name'].mean())
```

df['column_name']: Selects the specific column you want to fill.
.fillna(): The function used to replace missing values.
df['column_name'].median(): Calculates the median of the column and uses it as the fill value.

In [ ]:
# Replacing al NAs with other metrics:

df['down'] = df['down'].fillna(df['down'].median())

print(df.loc[:,'Date':'down'].head())


df['down'] = df['down'].fillna(df['down'].mean())

print(df.loc[:,'Date':'down'].head())


         Date      GameID  Drive  qtr  down
0  2009-09-10  2009091000      1    1   2.0
1  2009-09-10  2009091000      1    1   1.0
2  2009-09-10  2009091000      1    1   2.0
3  2009-09-10  2009091000      1    1   3.0
4  2009-09-10  2009091000      1    1   4.0
         Date      GameID  Drive  qtr  down
0  2009-09-10  2009091000      1    1   2.0
1  2009-09-10  2009091000      1    1   1.0
2  2009-09-10  2009091000      1    1   2.0
3  2009-09-10  2009091000      1    1   3.0
4  2009-09-10  2009091000      1    1   4.0


In [ ]:
# Mode is typically used for filling missing values in categorical columns. In this example column 'SideofField' contains
# strings (object data type)

print(df['SideofField'].dtype)

group_side_of_field = df.groupby('SideofField',dropna=False).size().sort_values(ascending=False)

print(group_side_of_field)

object
SideofField
OAK    13270
CLE    13212
BUF    13094
TEN    12968
MIA    12924
NYG    12855
ARI    12834
DET    12799
TB     12788
IND    12760
WAS    12674
PHI    12673
HOU    12628
DEN    12606
CHI    12583
NO     12566
KC     12486
SF     12452
BAL    12442
NYJ    12396
CIN    12384
DAL    12306
MIN    12277
GB     12250
CAR    12222
ATL    12088
SEA    12077
PIT    12023
NE     12001
SD     10952
JAC    10185
STL     9879
MID     4326
LA      2763
JAX     2754
LAC     1351
NaN      528
50       312
dtype: int64




```
grouped_side_of_field_with_nan = df.groupby('SideofField', dropna=False).size()
```





By default, groupby doesn't include missing values.
By adding dropna=False to the groupby() function explicitly instructs it to include missing values (NaN) as a separate group. This way, you'll see the count of missing values along with the counts for other unique values in the 'SideofField' column.

In [ ]:
df['SideofField'].isnull().sum()    #Checking if there NaN in the column 'SideofField'

528

In [ ]:
df['SideofField'] = df['SideofField'].fillna(df['SideofField'].mode()[0])   #Replacing missing values with the mode

df['SideofField'].isnull().sum()

print(df['SideofField'].mode())

0    OAK
Name: SideofField, dtype: object


In [ ]:
grouped_side_of_field_with_nan = df.groupby('SideofField', dropna=False).size()

print(grouped_side_of_field_with_nan.sort_values(ascending=False))

SideofField
OAK    13798
CLE    13212
BUF    13094
TEN    12968
MIA    12924
NYG    12855
ARI    12834
DET    12799
TB     12788
IND    12760
WAS    12674
PHI    12673
HOU    12628
DEN    12606
CHI    12583
NO     12566
KC     12486
SF     12452
BAL    12442
NYJ    12396
CIN    12384
DAL    12306
MIN    12277
GB     12250
CAR    12222
ATL    12088
SEA    12077
PIT    12023
NE     12001
SD     10952
JAC    10185
STL     9879
MID     4326
LA      2763
JAX     2754
LAC     1351
50       312
dtype: int64


Before applying the mode inputation, the mode of the column 'SideofField' is OAK with 13270 entries. After applying the mode inputation, the missing values were assigned to OAK adding 528 entries to OAK. The final result of OAK is 13798 entries.

Important Considerations:

1. Data Distribution: The choice of fill method (median, mean, mode) depends on the distribution of your data and the nature of the missing values.

2. Skewness: For skewed data, the median is often preferred over the mean as it's less sensitive to outliers.

3. Categorical Data: Mode is typically used for filling missing values in categorical columns.

4. Data Leakage: Be careful not to introduce data leakage when filling missing values. If you're using the data for machine learning, consider using imputation techniques within cross-validation folds.

# **The fillna(method='bfill') and fillna(method='ffill') methods in Pandas.**

fillna(method='bfill')

* What it does: This method propagates the next valid observation backward to fill the missing value. In simpler terms, it uses the value that comes after the missing value to fill the gap.
* How it works: bfill stands for "backward fill." It iterates through the DataFrame (or Series) and, when it encounters a missing value (NaN), it looks for the next valid (non-NaN) value in the same column. It then copies that next valid value into the missing value's place.
* Effect: This method is useful for filling missing values when you assume that the subsequent value is a reasonable approximation for the missing data. It can be helpful for time series data or situations where data is expected to be relatively continuous.

fillna(method='ffill')

* What it does: This method propagates the last valid observation forward to fill the missing value. In other words, it uses the value that comes before the missing value to fill the gap.
* How it works: ffill stands for "forward fill." It iterates through the DataFrame (or Series) and, when it encounters a missing value (NaN), it looks for the previous valid (non-NaN) value in the same column. It then copies that previous valid value into the missing value's place.
* Effect: This method is useful for filling missing values when you assume that the preceding value is a reasonable approximation for the missing data. It's often used for filling gaps in time series data or situations where data is expected to be relatively stable over time.

Important considerations:

Direction: bfill fills from the next valid value (backward), while ffill fills from the previous valid value (forward).
Data continuity: These methods assume some degree of continuity or stability in the data. If your data is highly volatile or unpredictable, using bfill or ffill might not be appropriate.
Potential bias: Be cautious when using these methods, as they can introduce bias into your data if the missing values are not randomly distributed.

In [ ]:
df['TimeSecs'].isnull().sum()

224

In [ ]:
# I want to know after applying bfill (or ffill) where the changes occured:
# Create a boolean mask indicating rows where 'TimeSecs' was originally null:  that's True for rows where 'TimeSecs' is
# null (NaN) and False otherwise. This mask stores the original null locations before applying bfill.

mask = df['TimeSecs'].isnull()

# that's True for rows where 'TimeSecs' is null (NaN) and False otherwise. This mask stores the original null
# locations before applying bfill.

In [ ]:
# Apply bfill to the 'TimeSecs' column

df['TimeSecs'] = df['TimeSecs'].fillna(method='bfill')

df['TimeSecs'].isnull().sum()

<ipython-input-23-d6bc4289ffa3>:3: FutureWarning: Series.fillna with 'method' is deprecated and will raise in a future version. Use obj.ffill() or obj.bfill() instead.
  df['TimeSecs'] = df['TimeSecs'].fillna(method='bfill')


0

In [ ]:
# Use the mask to select rows that were originally null. These are the rows where bfill potentially made changes.
changed_rows = df[mask]

# Display the 'TimeSecs' values in the changed rows
print(changed_rows['TimeSecs'])

2717       228.0
2719       219.0
4564      2700.0
4621      1800.0
4661       900.0
4723         0.0
7159      2700.0
7213      1800.0
7248       900.0
7304         0.0
11678     2700.0
11726     1800.0
16717     2145.0
18099     2700.0
18145     1800.0
18178      900.0
18222        0.0
28436     2700.0
28488     1800.0
28520      900.0
28578        0.0
30692     2700.0
30736     1800.0
30770      900.0
30817        0.0
35726       85.0
37163     2700.0
37244      900.0
37294        0.0
41920     2700.0
41973     1800.0
42009      900.0
42043        0.0
45916      900.0
53049      900.0
53095        0.0
55126      260.0
55383     1800.0
55425      900.0
55474        0.0
65835     2700.0
65890     1800.0
65928      900.0
65983        0.0
67615     2700.0
67711      900.0
67767        0.0
69505     1790.0
69515     1538.0
69517     1526.0
73686     1800.0
73724      900.0
73779        0.0
76600     1807.0
78801     2700.0
78845     1800.0
78877      900.0
78922        0.0
84521     2700

The above two lines of code are well explained here:



```
# changed_rows = df[mask]
```

This line selects rows from the DataFrame df where the mask mask is True. In other words, it extracts the rows that originally contained null values in the 'TimeSecs' column.

Think of the mask as a filter: it's used to select only those rows that meet the condition of having a True value in the mask (meaning they had a null in 'TimeSecs').
This selection occurs after the bfill has been applied, but the mask remembers the original locations of those null values.

```
# print(changed_rows['TimeSecs'])
```
This line simply prints the 'TimeSecs' values for the rows we extracted in the previous step (the changed_rows DataFrame). These values now reflect what the bfill method filled them with.

In summary:

The mask stores the locations of the original null values in the 'TimeSecs' column.
After applying bfill, we use the mask to extract the rows where those nulls were present.
By printing 'TimeSecs' for those extracted rows, we see the values that replaced the original nulls.

The output is telling you:

"In the original DataFrame df, the row (e.g. first row of the ouput) with index 2717 had a missing value (NaN) in the 'TimeSecs' column. After applying the bfill method, this missing value was replaced with the value 228.0."

In [ ]:
df[df['desc'].isnull()]   # this query shows the rows with values NaN in the columns 'desc'

Date      GameID  Drive  qtr  down   time  TimeUnder  TimeSecs  \
93350   2011-09-18  2011091805     19    4   2.0  05:17          6     317.0   
109947  2011-11-06  2011110600     23    4   2.0  03:14          4     194.0   

        PlayTimeDiff SideofField  yrdln  yrdline100  ydstogo  ydsnet  \
93350            6.0         ARI    2.0         2.0        0      73   
109947           0.0         BUF   35.0        65.0        0       0   

        GoalToGo  FirstDown posteam DefensiveTeam desc  PlayAttempted  \
93350        0.0        0.0     WAS           ARI  NaN              1   
109947       0.0        0.0     BUF           NYJ  NaN              1   

        Yards.Gained  sp  Touchdown ExPointResult TwoPointConv DefTwoPoint  \
93350              0   0          0           NaN          NaN         NaN   
109947             0   0          0           NaN          NaN         NaN   

        Safety  Onsidekick PuntResult PlayType Passer Passer_ID  PassAttempt  \
93350        0           0        NaN      Run    NaN       NaN            0   
109947       0           0        NaN      Run    NaN       NaN            0   

       PassOutcome PassLength  AirYards  YardsAfterCatch  QBHit PassLocation  \
93350          NaN        NaN         0                0      0          NaN   
109947         NaN        NaN         0                0      0          NaN   

        InterceptionThrown Interceptor Rusher Rusher_ID  RushAttempt  \
93350                    0         NaN    NaN       NaN            1   
109947                   0         NaN    NaN       NaN            1   

       RunLocation RunGap Receiver Receiver_ID  Reception ReturnResult  \
93350          NaN    NaN      NaN         NaN          0          NaN   
109947         NaN    NaN      NaN         NaN          0          NaN   

       Returner BlockingPlayer Tackler1 Tackler2 FieldGoalResult  \
93350       NaN            NaN      NaN      NaN             NaN   
109947      NaN            NaN      NaN      NaN             NaN   

        FieldGoalDistance  Fumble RecFumbTeam RecFumbPlayer  Sack  \
93350                 NaN       0         NaN           NaN     0   
109947                NaN       0         NaN           NaN     0   

        Challenge.Replay ChalReplayResult  Accepted.Penalty PenalizedTeam  \
93350                  0              NaN                 0           NaN   
109947                 0              NaN                 0           NaN   

       PenaltyType PenalizedPlayer  Penalty.Yards  PosTeamScore  DefTeamScore  \
93350          NaN             NaN              0          19.0          21.0   
109947         NaN             NaN              0          11.0          27.0   

        ScoreDiff  AbsScoreDiff HomeTeam AwayTeam  Timeout_Indicator  \
93350        -2.0           2.0      WAS      ARI                  0   
109947      -16.0          16.0      BUF      NYJ                  0   

       Timeout_Team  posteam_timeouts_pre  HomeTimeouts_Remaining_Pre  \
93350           NaN                     3                           3   
109947          NaN                     0                           0   

        AwayTimeouts_Remaining_Pre  HomeTimeouts_Remaining_Post  \
93350                            2                            3   
109947                           3                            0   

        AwayTimeouts_Remaining_Post  No_Score_Prob  Opp_Field_Goal_Prob  \
93350                             2            NaN                  NaN   
109947                            3            NaN                  NaN   

        Opp_Safety_Prob  Opp_Touchdown_Prob  Field_Goal_Prob  Safety_Prob  \
93350               NaN                 NaN              NaN          NaN   
109947              NaN                 NaN              NaN          NaN   

        Touchdown_Prob  ExPoint_Prob  TwoPoint_Prob  ExpPts  EPA  airEPA  \
93350              NaN           0.0            0.0     NaN  NaN     NaN   
109947             NaN           0.0     

In [ ]:
df['desc'].dtypes   #The ouput indicates there different datatypes in this column (strings and NaNs)

dtype('O')

# **More advanced methods to handle missing values**

  1.K-Nearest Neighbors (KNN) Imputation:
  Concept: This method leverages the idea that similar data points tend to have similar values. It identifies 'k' nearest neighbors to a data point with missing values and uses the average (or weighted average) of their values to fill the gap.**It expects numerical data to perform these calculations.**


  Advantages: Can handle both numerical and categorical data, often provides more accurate imputations than simple methods.

  Disadvantages: KNNImputer cannot directly handle string or categorical data.

  Considerations: Computationally expensive for large datasets, sensitive to the choice of 'k' and distance metric.

  While imputing a single column might be sufficient in some cases, remember that KNN imputation considers the relationships between features. Imputing multiple related columns together can potentially lead to more accurate results.

In [ ]:
import kagglehub
import pandas as pd
from sklearn.impute import KNNImputer

# Download latest version
path = kagglehub.dataset_download("maxhorowitz/nflplaybyplay2009to2016")

df = pd.read_csv(path + "/NFL Play by Play 2009-2017 (v4).csv")

# 1. Create a KNNImputer instance
imputer = KNNImputer(n_neighbors=3)  # You can adjust the number of neighbors

# 2. Reshape the 'down' column to a 2D array (required by KNNImputer)
down_values = df['down'].values.reshape(-1, 1)

# 3. Apply the imputer to the 'down' column
imputed_down = imputer.fit_transform(down_values)  # Execution took 15 minutes!

# 4. Replace the original 'down' column with the imputed values
df['down'] = imputed_down

In [ ]:
columns_data = df['down'].isnull().sum()

print(f"number of null values after applying KNNImputer: {columns_data}")

print(df['down'])

In this scenario, I have applied KNNImputer on a single column. This is how it works:

1. Limited Information: Since KNNImputer relies on distances between data points to identify neighbors, when applied to only one column, it has limited information to work with. It essentially treats each value in that column as a single-dimensional data point.
2. Distances in 1D: The distances between data points are calculated based solely on the values in the specified column. In your case, the distances between rows would be the absolute difference between their 'down' values.
3. Neighbor Selection: The KNNImputer still tries to find the 5 nearest neighbors (based on these 1D distances) for each missing value in the 'down' column.
Imputation: The missing 'down' value is then imputed using the average of the 'down' values from those 5 nearest neighbors.

In essence, applying KNNImputer to a single column performs a localized imputation based on the values in that column alone. While it can be a valid approach in some cases, it's crucial to be aware of its limitations and consider the broader context of your dataset.